### PR TITLE
Ignore line endings in exclude-file

### LINKED ISSUE #1888: Add a way to ignore line endings when matching lines from `--exclude-file`
This can be especially when repository is used on Windows with git's `core.autocrlf` set to `true` as then some files can have CRLF line endings locally even though they're checked out into git with LF endings. This mixing can therefore cause `codespell` to not ignore some lines if the line endings in `--exclude-file` differ from line endings in the file with spelling issue.

### PR DESCRIPTION
**Pull request by Jackenmen**
Fixes #1888

### CODE REVIEW

File: codespell_lib/tests/test_basic.py
Comment (peternewman) [inline codespell_lib/tests/test_basic.py:253]: You're only checking for \r\n now, which means it would be possible for us to make a change which requires this exclusively. Can you make one of them \n and one \r\n to cover both cases?

Actually I guess for a good test here we should have four lines in bad.txt and four in tmp.txt so we can check all combinations work.

File: codespell_lib/_codespell.py
Comment (peternewman) [inline codespell_lib/_codespell.py:419]: This will remove any whitespace, not just the line ending. I'm not sure if that's necessarily an issue, but it's a bigger change than the minimum needed to fix this, so we should possibly have some whitespace test cases too.

**Review by peternewman — CHANGES_REQUESTED** [review]
Can you extend test_ignore_dictionary too, so that also covers both cases?

**Comment by jack1142** [comment]
Alright, I added the test cases for all combinations of these line endings `\n`, `\r\n`, ` \n`, ` \r\n`.

It seems that I need to use different misspellings for `test_ignore_dictionary` as the ones I chose don't seem to be detected as misspellings and therefore the test cases don't quite work. Got any suggestions on what words I could use here?

**Comment by jack1142** [comment]
> Got any suggestions on what words I could use here?
Nevermind, I found some in the other tests 😄

File: codespell_lib/tests/test_basic.py
Comment (peternewman) [inline codespell_lib/tests/test_basic.py:226]: It would probably be good to mix in some \r\n in here too.
```suggestion
        f.write('1 abandonned 1\n'
                '2 abandonned 2\n'
                '3 abandonned 3\r\n'
                '4 abilty 4\n'
                '5 abilty 5\n'
                '6 abilty 6\r\n'
                '7 ackward 7\n'
                '8 ackward 8\n'
                '9 ackward 9\r\n'
                'abondon\n')
```

File: codespell_lib/_codespell.py
Comment (peternewman) [inline codespell_lib/_codespell.py:605]: I've realised the issue. This will write back out. See the `-w` based tests (and extend them to prove it/stop us breaking it anyway).

An input of `abandonned\nAbandonned\r\nABANDONNED \nAbAnDoNnEd` will become `abandonned\nAbandonned\nABANDONNED\nAbAnDoNnEd` I think

**Review by peternewman — CHANGES_REQUESTED** [review]
Sorry the write changes thing will need fixing somehow. Perhaps just move it to when you're comparing the line to the exclude list?

File: codespell_lib/_codespell.py
Comment (jack1142) [reply codespell_lib/_codespell.py:605]: > An input of `abandonned\nAbandonned\r\nABANDONNED \nAbAnDoNnEd` will become `abandonned\nAbandonned\nABANDONNED\nAbAnDoNnEd` I think

It becomes `abandoned\nAbandoned\nABANDONED \nabandoned` on both codespell 2.0.0 and on the version with modifications from this PR.
I'm not quite sure what and why I should be extending tests here as the changes from this PR do not make any changes that would cause any regression. I feel like it would be better suited for a separate PR?

If you don't agree, I can make the changes here, just let me know *what test* I should be modifying and/or what kind of inputs/outputs should I be checking there.

Comment (jack1142) [reply codespell_lib/_codespell.py:605]: @peternewman do you think I could get your decision here so that I know if this is fine or if I should work on something before this will be able to move forward?

Comment (tenshiAMD) [reply codespell_lib/_codespell.py:605]: ping @peternewman

Comment (peternewman) [reply codespell_lib/_codespell.py:605]: > > An input of `abandonned\nAbandonned\r\nABANDONNED \nAbAnDoNnEd` will become `abandonned\nAbandonned\nABANDONNED\nAbAnDoNnEd` I think
> 
> It becomes `abandoned\nAbandoned\nABANDONED \nabandoned` on both codespell 2.0.0 and on the version with modifications from this PR.

Odd, as we added the tests here in #2490 and they seem to behave okay, although admittedly they are a slightly different code path to the command line version. What OS were you testing on?

> I'm not quite sure what and why I should be extending tests here as the changes from this PR do not make any changes that would cause any regression. I feel like it would be better suited for a separate PR?

I believe the change you're making will introduce a bug, or perhaps simply further embed an existing bug, so I figure fixing the crucial underlying one first is probably a good idea.

> If you don't agree, I can make the changes here, just let me know _what test_ I should be modifying and/or what kind of inputs/outputs should I be checking there.

As mentioned, I think we've now done that in #2490 so when that's in, you can just update/rebase and as long as your code continue to pass we'll be all good.

Comment (jack1142) [reply codespell_lib/_codespell.py:605]: > What OS were you testing on?

Back then probably only on Windows 10.

> I believe the change you're making will introduce a bug, or perhaps simply further embed an existing bug, so I figure fixing the crucial underlying one first is probably a good idea.

That certainly makes sense, I am just not really sure what the issue is exactly. Hopefully, the tests added in #2490 will make it more clear whether there is any issue and if so, what that issue is, thanks for working on those tests. Once that PR gets merged, I'll rebase.
If any further work on the PR will be need, I might not get to it before Saturday but other than that, I'll be happy to make any necessary changes :+1:

Comment (jack1142) [reply codespell_lib/_codespell.py:642]: I rebased the PR on top of current `master` now that the tests got merged. Locally the tests ran fine, you'll have to approve the CI run though since I haven't had any PRs into this repository yet.

---

By the way, in case you don't know, the requirement for approval of workflows for first-time contributors who are NOT new to GitHub can be disabled to decrease the contribution barrier for first-time contributors, see:
https://twitter.com/JakubKuczys/status/1562660410108841986

**Comment by luzpaz** [comment]
@peternewman care to review this?

**Comment by Jackenmen** [comment]
I resolved conflicts in case anyone wants to look at this again.

**Comment by Jackenmen** [comment]
If this one ever gets merged, I *may* look into it.

**Review by DimitriPapadopoulos — APPROVED** [review]

**Comment by DimitriPapadopoulos** [comment]
@peternewman Would you agree to merge this?

**Comment by Andy-Grigg** [comment]
Is there any interest in getting this merged? As far as I can tell, the current implementation cannot work at all with Windows line endings in the exclude-lines file

`open()` is called in `build_exclude_hashes` without the `newline` parameter, which normalizes all line endings to `\n`, but everywhere else `open()` is called with `newline=''`, so newlines are preserved.

As a result, if there's a `\r\n` ending in the file to be checked, it's impossible to match any line to the exclude-lines file.

**Comment by Jackenmen** [comment]
This PR is supposed to cause line endings in spell-checked files to be ignored when being compared to entries in exclude lines file. This feature is wanted and I definitely have interest in getting this PR merged. You are the first to say that the current implementation has issues (more specifically, you claim that it doesn't work) so there was no way for me to know there are some concerns about it before you posted your comment :p

I believe the implementation is sound though - the lines are rstripped before checking so the way files are opened shouldn't matter. Test cases also seem to indicate that, are you able to provide reproduction steps for which the changes in this PR aren't working as intended?

**Comment by Andy-Grigg** [comment]
> This PR is supposed to cause line endings in spell-checked files to be ignored when being compared to entries in exclude lines file. This feature is wanted and I definitely have interest in getting this PR merged. You are the first to say that the current implementation has issues (more specifically, you claim that it doesn't work) so there was no way for me to know there are some concerns about it before you posted your comment :p
> 
> I believe the implementation is sound though - the lines are rstripped before checking so the way files are opened shouldn't matter. Test cases also seem to indicate that, are you able to provide reproduction steps for which the changes in this PR aren't working as intended?

I'm sorry, I wasn't clear in my comment. This work is brilliant and I love it, I just checked out your branch and the ignore-lines functionality works as expected. As you say, since you're stripping off the whitespace, it doesn't matter how `open()` is set up.

My comment was addressed more to the maintainer - I was really just trying to add my vote for getting this merged.

**Comment by DimitriPapadopoulos** [comment]
@Jackenmen Could you rebase one last time, as there are conflicts? I'll merge when you're done.

**Comment by Jackenmen** [comment]
> Could you rebase one last time, as there are conflicts?

@DimitriPapadopoulos It doesn't seem that there were any conflicts, I just made a regular merge commit through GitHub UI. Do let me know if you're not using squash merges and need me to rebase instead to keep a clean history, I can do that instead :smile:

**Comment by Jackenmen** [comment]
> 'Current implementation' in my comment was referring to the released functionality, not the changes that you've done here!

Makes sense, both our interpretations of your comment were definitely valid, just the wonders of misunderstandings in text-based communication :) I guess I just assumed it's about my PR since that's where your comment was posted + you also went into details of the implementation (but as it turns out, they were details of the implementation in `master` *as well*). No worries though, I'm just glad you like it!

**Comment by DimitriPapadopoulos** [comment]
@Jackenmen Yes, I was thinking of squash merging.

**Comment by Jackenmen** [comment]
No need for me to do anything else then, great! Excited to see this finally get into the project!

**Comment by DimitriPapadopoulos** [comment]
@Jackenmen Thank you for your contribution and your patience.

**Pull request merged**
